src/string/extglobmatch.c

```diff
@@ -0,0 +1,276 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "extglobmatch.h"
+#include "brz.h"
+
+#define cat_one()    do{                                    \
+                        if (atoms > 1) {                    \
+                            --atoms;                        \
+                            glob *b = *--frag;              \
+                            glob *a = *--frag;              \
+                            *frag++ = glob_seq(a,b);             \
+                            free_glob(a);                   \
+                            free_glob(b);                   \
+                        }                                   \
+                    } while(0)
+
+#define cat_all()    do{                                    \
+                          while (--atoms > 0) {             \
+                              glob *b = *--frag;            \
+                              glob *a = *--frag;            \
+                              *frag++ = glob_seq(a,b);           \
+                              free_glob(a);                 \
+                              free_glob(b);                 \
+                          }                                 \
+                    }while(0)
+
+static glob* match_glob = NULL;
+
+static void cleanup(glob **start, glob **end) {
+    while (start != end)
+    {
+        free_glob(*start);
+        ++start;
+    }
+}
+
+static void print_error(char const c) {
+    fprintf(stderr, "pattern error: Unexpected character '%c'\n", c);
+}
+
+static glob* handle_operator(char op, glob const* g)
+{
+    glob* tmp = NULL;
+    switch (op) {
+        case '!': tmp = glob_not(g);
+                  break;
+        case '*': tmp = glob_star(g);
+                  break;
+        case '+': tmp = glob_plus(g);
+                  break;
+        case '@': tmp = glob_at(g);
+                  break;
+        case '?': tmp = glob_opt(g);
+                  break;
+    }
+    return tmp;
```

Comment:
Uh, let me rephrase that actually: add an `assert(tmp);` because tmp being NULL is not possible.